### ISSUE TITLE
About the conv4d function

### ISSUE THREAD

**Issue by monaen** [mhuen/TFScripts#1]
Hi there, 
Your 4d convolution implementation is a great help to me. actually I also implement a pytorch version 4d conv but the calculation takes too much time. Can you tell me how do you think of 4d convolution problem as an extension of 3d convolution. I want to know how to extend the 3d conv operation to 4d conv operation.  And also is this implementation of 4d convolution equivalent to traditional 4d convolution? Thanks!!!

**Comment by mhuen** [comment]
Hi,
the way I do it, is that I use tensorflow's 3d convolution to convolve three of the four dimensions. Only the fourth dimension is plugged together on python level. Using tensorflow's 3d convolution is a lot faster than plugging the other three dimensions together on python level as well.

If you look at the math, you can see that you can split up the 4d convolution operator, by performing a 3d convolution for every slice in the fourth dimension (let's just say the t-dimension) and then to add up the results manually, to incorporate the t-dimension.

So, for every output t-value t_i, e.g. [ : , : , : , :, t_i, :] with the dimensions [batch, x-dim, y-dim, z-dim, t-dim, output channel], I collect all the necessary slices [:, :, :, :, t_k, :] ([batch, x-dim, y-dim, z-dim, t-dim, input channel]) of the input data. For a filter of size 3 in the t-dimension, I would need to collect three slices. 
For example, if t_i is 10 and my filter in the t-dimension has the size of 3, I would collect the slices:
[:, :, :, :, t_k=9, :]
[:, :, :, :, t_k=10, :]
[:, :, :, :, t_k=11, :]
of my input data.

I then convolve these slices with the corresponding kernel via tensorflow's 3d convolution to handle the other three dimensions.
For example:

result_for_t_k=9 : conv3d( InputData[:, :, :, :, t_k=9, :], Kernel[:, :, :, 0,:, :]) 
result_for_t_k=10 : conv3d( InputData[:, :, :, :, t_k=10, :], Kernel[:, :, :, 1,:, :]) 
result_for_t_k=11 : conv3d( InputData[:, :, :, :, t_k=11, :], Kernel[:, :, :, 2,:, :]) 

where the dimensions for the Kernel are: [ x-dim, y-dim, z-dim, t-dim, input channels, output channels].

After this step I need to add the results of the slices together:

result_t_i=10 = sum ( result_for_t_k=9, result_for_t_k=10, result_for_t_k=11)

The result is the output value for Output[ : , : , : , :, t_i=10, :].
This has to be done for every t_i to obtain the complete output.


And that is essentially how the function conv4d_stacked works. The only difference is that you can choose over which dimension you want to stack and over which you want to perform tensorflow's 3d convolution. And of course, it handles things like padding and dilation, which needs to be considered when putting together the slices. I hope that made it a bit clearer on how I put that together. 



I'm not quite sure I understand what you mean with "traditional 4d convolution". For one thing, tensorflow's convolution operators actually perform a cross-correlation rather than a convolution, as also mentioned in the documenation: https://www.tensorflow.org/api_docs/python/tf/nn/convolution .
Although this should essentially be the same, since the weights are learned anyway, so it doesn't matter if the kernel is flipped first. But if you mean with "traditional 4d convolution" the convolution as it is also used for 3d and 2d convolution in tensorflow, then yes, this is a "traditional 4d convolution". If tensorflow were to add an official conv4d operation, it would produce the same results as this implementation( apart from maybe some conventional choices). I verified this by doing the same with a conv3d_stacked function which stacks along the z-dim and uses tensorflow's conv2d for the other two dimensions. It produced the same results, except for some combinations of stride size and VALID pattern I think, which I'm pretty sure is due to some conventional choices.
Some conventional choices could  for instance be:
If my kernel has an even size, do I include a pixel more on the left or on the right in the convolution?


That said, a 4d convolution will not always be the best option and it will most certainly greatly increase the computational complexity. The naming scheme I used is calling the fourth dimension the time dimension. However, especially with the time dimension and time series, you might want to look into other things like LSTMs for example.

Let me know if this answers your questions.

**Comment by monaen** [comment]
Thank you so much for the detailed explanation !!! And yes, you have answered my questions. My current research is something about light field image analysis, and the fourth dimension is not time dimension, so I think 4d convolution is more reasonable. 

Thanks again...
